src/libraries/System.Security.Cryptography/tests/KmacTestDriver.cs

```diff
@@ -0,0 +1,1024 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.XUnitExtensions;
+using Xunit;
+
+namespace System.Security.Cryptography.Tests
+{
+    public interface IKmacTrait<TKmac> where TKmac : IDisposable
+    {
+        static abstract TKmac Create(ReadOnlySpan<byte> key, ReadOnlySpan<byte> customizationString);
+        static abstract TKmac Create(byte[] key, byte[] customizationString);
+        static abstract bool IsSupported { get; }
+        static abstract void AppendData(TKmac kmac, byte[] data);
+        static abstract void AppendData(TKmac kmac, ReadOnlySpan<byte> data);
+        static abstract byte[] GetHashAndReset(TKmac kmac, int outputLength);
+        static abstract void GetHashAndReset(TKmac kmac, Span<byte> destination);
+        static abstract byte[] GetCurrentHash(TKmac kmac, int outputLength);
+        static abstract void GetCurrentHash(TKmac kmac, Span<byte> destination);
+
+        static abstract byte[] HashData(byte[] key, byte[] source, int outputLength, byte[] customizationString);
+        static abstract byte[] HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, int outputLength, ReadOnlySpan<byte> customizationString);
+        static abstract void HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, Span<byte> destination, ReadOnlySpan<byte> customizationString);
+
+        static abstract byte[] HashData(byte[] key, Stream source, int outputLength, byte[] customizationString);
+        static abstract byte[] HashData(ReadOnlySpan<byte> key, Stream source, int outputLength, ReadOnlySpan<byte> customizationString);
+        static abstract void HashData(ReadOnlySpan<byte> key, Stream source, Span<byte> destination, ReadOnlySpan<byte> customizationString);
+
+        static abstract ValueTask HashDataAsync(
+            ReadOnlyMemory<byte> key,
+            Stream source,
+            Memory<byte> destination,
+            ReadOnlyMemory<byte> customizationString,
+            CancellationToken cancellationToken);
+
+        static abstract ValueTask<byte[]> HashDataAsync(
+            ReadOnlyMemory<byte> key,
+            Stream source,
+            int outputLength,
+            ReadOnlyMemory<byte> customizationString,
+            CancellationToken cancellationToken);
+
+        static abstract ValueTask<byte[]> HashDataAsync(
+            byte[] key,
+            Stream source,
+            int outputLength,
+            byte[] customizationString,
+            CancellationToken cancellationToken);
+    }
+
+    public abstract class KmacTestDriver<TKmacTrait, TKmac>
+        where TKmacTrait : IKmacTrait<TKmac>
+        where TKmac : IDisposable
+    {
```

Comment:
I think you've already done this once before, but... "wow, I like how nicely this works"

Reply:
Thanks! Yep this is very close to what I did for SHAKE.